tensorboard/webapp/plugins/npmi/views/violin_filters/violin_filter/violin_filter_component.ts

```diff
@@ -0,0 +1,413 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {
+  ChangeDetectionStrategy,
+  Component,
+  Input,
+  Output,
+  AfterViewInit,
+  OnChanges,
+  SimpleChanges,
+  EventEmitter,
+  ViewChild,
+  ElementRef,
+} from '@angular/core';
+
+import * as d3 from 'd3';
+
+import {MetricFilter} from './../../../store/npmi_types';
+import {ViolinChartData, ViolinBin} from './../../../util/violin_data';
+
+@Component({
+  selector: 'violin-filter-component',
+  templateUrl: './violin_filter_component.ng.html',
+  styleUrls: ['./violin_filter_component.css'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ViolinFilterComponent implements AfterViewInit, OnChanges {
+  @Input() metricName!: string;
+  @Input() filter!: MetricFilter;
+  @Input() chartData!: {
+    violinData: ViolinChartData;
+    extremes: {min: number; max: number};
+  };
+  @Input() width!: number;
+  @Output() onRemove = new EventEmitter();
+  @Output() onUpdateFilter = new EventEmitter<MetricFilter>();
+  @ViewChild('chart', {static: true, read: ElementRef})
+  private readonly chartContainer!: ElementRef<HTMLDivElement>;
+  private height = 300;
+  private margin = {top: 20, right: 10, bottom: 20, left: 10};
+  private drawMargin = {top: 0, right: 0, bottom: 20, left: 20};
+  get chartWidth(): number {
+    return this.width - this.margin.left - this.margin.right;
+  }
+  get chartHeight(): number {
+    return this.height - this.margin.top - this.margin.bottom;
+  }
+  get drawHeight(): number {
+    return this.chartHeight - this.drawMargin.top - this.drawMargin.bottom;
+  }
+  get drawWidth(): number {
+    return this.chartWidth - this.drawMargin.left - this.drawMargin.right;
+  }
+  // Drawing containers
+  private svg: any;
```

Comment:
Please type them more specifically. Is this `d3.Selection<SvgElement>`?

I understand that d3 types are cumbersome to type but I really think `any`s will greatly impair readability.